plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java

```diff
@@ -42,11 +62,55 @@ protected void validate(IResource resource, byte[] bytes)
       Map<BannedElement, Integer> bannedElementOffsetMap =
           ValidationUtils.getOffsetMap(bytes, parserResults);
       for (Map.Entry<BannedElement, Integer> entry : bannedElementOffsetMap.entrySet()) {
-        createMarker(resource, entry.getKey(), entry.getValue());
+        String servletClassMarker =
+            "com.google.cloud.tools.eclipse.appengine.validation.undefinedServletMarker";
+        if (servletClassMarker.equals(entry.getKey().getMarkerId())) {
+          UndefinedServletElement element = (UndefinedServletElement) entry.getKey();
+          String servletClass = element.getServletClass();
+          if (findClass(servletClass, resource.getProject()) == null) {
+            createMarker(resource, entry.getKey(), entry.getValue());
+          }
+        } else {
+          createMarker(resource, entry.getKey(), entry.getValue());
+        }
       }
     } catch (SAXException ex) {
       createSaxErrorMessage(resource, ex);
     }
   }
+  
+  /**
+   * Checks for the given class name within the project.
+   */
+  static IType findClass(String typeName, IProject project) {
+    if (typeName == null || "".equals(typeName)) {
+      return null;
+    }
+    SearchPattern pattern = SearchPattern.createPattern(typeName, 
+        IJavaSearchConstants.CLASS,
+        IJavaSearchConstants.DECLARATIONS,
+        SearchPattern.R_EXACT_MATCH | SearchPattern.R_ERASURE_MATCH);
+    IJavaSearchScope scope = SearchEngine.createJavaSearchScope(
+        new IJavaElement[] { (IJavaElement) JavaCore.create(project) });
+    CollectingSearchRequestor requestor = new CollectingSearchRequestor();
+    performSearch(pattern, scope, requestor, null);
+    List<SearchMatch> results = requestor.getResults();
+    return results.isEmpty() ? null : (IType) results.get(0).getElement();
+  }
+  
+  /**
+   * Performs the search for a class of a given pattern within the project.
+   */
+  static void performSearch(SearchPattern pattern, IJavaSearchScope scope, SearchRequestor requestor,
+      IProgressMonitor monitor) {
+    SearchEngine searchEngine = new SearchEngine();
```

Comment:
Do we need/want to create a new one of these each time? If it's not expensive, fine; but if this is an expensive and thread-safe object then maybe create once in a field.